File: [scripts]/al_mdt/client/nui/modules/reports.lua
RegisterNUICallback('addRecordEntry', function(data, cb)
    TriggerServerEvent('al_mdt:AddReport', mdt.k, data.identifier, data.entry.type, data.entry.submittedBy, data.entry.description, json.encode(data.entry.input), data.entry.data, data.entry.date)

    cb({})
end)

RegisterNUICallback('updateRecordEntry', function(entry, cb)
    TriggerServerEvent('al_mdt:UpdateReport', mdt.k, entry.id, entry.type, entry.description, json.encode(entry.input), json.encode(entry.data), entry.date)

    cb({})
end)

RegisterNUICallback('deleteRecordEntry', function(entryId, cb)
    TriggerServerEvent('al_mdt:DeleteReport', mdt.k, entryId)

    cb({})
end)

RegisterNUICallback('fetchReport', function(id, cb)
    TriggerServerCallback('al_mdt:FetchReport', function(resp)
        if resp then
            cb(resp)
        else
            cb(false)
        end
    end, mdt.k, id)
end)

RegisterNUICallback('fetchSelfRecentReports', function(identifier, cb)
    TriggerServerCallback('al_mdt:FetchRecentReports', function(resp)
        cb(resp)
    end, mdt.k, identifier)
end)

RegisterNUICallback('fetchReports', function(data, cb)
    TriggerServerCallback('al_mdt:FetchReports', function(resp)
        local pageData = resp.pageData

        if pageData then
            for i=1, #pageData do
                local report = pageData[i]
                report.input = json.decode(report.input)
                report.submittedBy = report.input.submittedBy
                report.input = nil
            end
        else
            pageData = {}
        end

        cb(resp)
    end, mdt.k, data.page, data.getTotalPages, data.filter)
end)

RegisterNUICallback('jailPlayer', function(data, cb)
    if data.time <= 0 then return end
    TriggerServerEvent('al_mdt:JailPlayer', mdt.k, data.identifier, data.reason, data.time)

    cb({})
end)

File: [scripts]/al_mdt/ui/src/api/modules/reports.js
import store from '@/store'
const resourceName = 'al_mdt'

class reportsApi {
    async Post(method, data) {
        if (process.env.NODE_ENV !== 'development') {
            const formattedData = data === undefined ? '{}' : JSON.stringify(data)

            try {
                let response = await fetch(`https://${resourceName}/${method}`, {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json; charset=UTF-8',
                    },
                    body: formattedData
                });

                return await response.json()
            } catch (error) {
                return false
            }
        }
    }

    // [POST] Jail Player
    async JailPlayer(identifier, reason, time) {
        try {
            this.Post('jailPlayer', {
                identifier: identifier,
                reason: reason,
                time: time,
            })
        } catch (error) {
            console.log('Error caught (Jailing player): ', error)
            return false
        }
    }

    // [POST] Fetch report
    async FetchReport(id) {
        try {
            return await this.Post('fetchReport', id)
        } catch (error) {
            console.log('Error caught (Fetching report): ', error)
            return false
        }
    }

    // [POST] Fetch criminal record
    async FetchReportHistory(identifier) {
        return await this.Post('getCriminalRecord', identifier)
    }

    // [POST] Fetch recent reports
    async FetchRecentReports() {
        try {
            return await this.Post('fetchSelfRecentReports', store.getters.GetSelf().identifier)
        } catch (error) {
            console.log('Error caught (Fetching self recent reports): ', error)
            return false
        }
    }

    // [POST] Fetch all reports
    async FetchAllReports(page, getTotalPages, filter) {
        try {
            return await this.Post('fetchReports', {
                page: page,
                getTotalPages: getTotalPages,
                filter: filter
            })
        } catch (error) {
            console.log('Error caught (Fetching all reports): ', error)
            return false
        }
    }

    // [POST] Add (incident or arrest) report entry
    AddRecordEntry(entry) {
        try {
            this.Post('addRecordEntry', entry)
        } catch (error) {
            console.log('Error caught (Adding report): ', error)
        }
    }

    // [POST] Update (incident or arrest) report entry
    UpdateRecordEntry(entry) {
        try {
            this.Post('updateRecordEntry', entry)
        } catch (error) {
            console.log('Error caught (Updating report): ', error)
        }
    }

    // [POST] Delete (incident or arrest) report entry
    DeleteRecordEntry(entryId) {
        try {
            this.Post('deleteRecordEntry', entryId)
        } catch (error) {
            console.log('Error caught (Deleting report): ', error)
        }
    }

    // [POST] Fetch fines
    async FetchFines(identifier) {
        try {
            return await this.Post('fetchFines', identifier)
        } catch (error) {
            console.log('Error caught (Fetching fines): ', error)
        }
    }

    // [POST] Update fine
    async PayFine(data) {
        try {
            return await this.Post('payFine', data)
        } catch (error) {
            console.log('Error caught (Updating fines): ', error)
        }
    }

    // [POST] Exempt fine
    ExemptFine(id) {
        try {
            this.Post('exemptFine', id)
        } catch (error) {
            console.log('Error caught (Exempting fine): ', error)
        }
    }
}

const instance = new reportsApi()

export default instance